nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ProcessGroupResource.java

```diff
@@ -2836,6 +2950,331 @@ public Response importProcessGroup(
+
+                    // prune response as necessary
+                    for (ProcessGroupEntity childGroupEntity : pasteResponseEntity.getFlow().getProcessGroups()) {
+                        childGroupEntity.getComponent().setContents(null);
+                    }
+
+                    // create the response entity
+                    populateRemainingSnippetContent(pasteResponseEntity.getFlow());
+
+                    return generateOkResponse(pasteResponseEntity).build();
+                }
+        );
+    }
+
+    private static VersionedProcessGroup getVersionedProcessGroup(final CopyResponseEntity copyResponse) {
+        final VersionedProcessGroup versionedProcessGroup = new VersionedProcessGroup();
+        versionedProcessGroup.setProcessors(new HashSet<>(copyResponse.getProcessors()));
+        versionedProcessGroup.setInputPorts(new HashSet<>(copyResponse.getInputPorts()));
+        versionedProcessGroup.setOutputPorts(new HashSet<>(copyResponse.getOutputPorts()));
+        versionedProcessGroup.setProcessGroups(new HashSet<>(copyResponse.getProcessGroups()));
+        versionedProcessGroup.setRemoteProcessGroups(new HashSet<>(copyResponse.getRemoteProcessGroups()));
+        versionedProcessGroup.setFunnels(new HashSet<>(copyResponse.getFunnels()));
+        versionedProcessGroup.setLabels(new HashSet<>(copyResponse.getLabels()));
+        versionedProcessGroup.setConnections(new HashSet<>(copyResponse.getConnections()));
+        return versionedProcessGroup;
+    }
+
+    private void mapVersionedIds(final VersionedProcessGroup group, final Map<String, String> idMapping, final Map<String, String> serviceIdMapping) {
+        group.getControllerServices().forEach(cs -> {
+            final String newId = generateUuid(cs.getIdentifier());
+            idMapping.put(cs.getIdentifier(), newId);
+            serviceIdMapping.put(cs.getIdentifier(), newId);
+            cs.setIdentifier(newId);
+        });
+        group.getControllerServices().forEach(cs -> {
+            cs.getProperties().entrySet().stream()
+                    .filter(propertyEntry -> {
+                        final Map<String, VersionedPropertyDescriptor> propertyDescriptors = cs.getPropertyDescriptors();
+                        if (propertyDescriptors != null) {
+                            final VersionedPropertyDescriptor propertyDescriptor = propertyDescriptors.get(propertyEntry.getKey());
+                            if (propertyDescriptor != null && propertyDescriptor.getIdentifiesControllerService()) {
+                                return serviceIdMapping.containsKey(propertyEntry.getValue());
+                            }
+                        }
+
+                        return false;
+                    })
+                    .findFirst()
+                    .ifPresent(serviceEntry -> serviceEntry.setValue(serviceIdMapping.get(serviceEntry.getValue())));
```

Comment:
I think we want to use `.forEach` here instead of `.findFirst().ifPresent()` because we want to do this for all properties, right?